ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java

```diff
@@ -0,0 +1,593 @@
+    if (operationException.get() != null) {
+      throw operationException.get();
+    }
+    if (readCompleteException.get() != null) {
+      throw readCompleteException.get();
+    }
+    Assert.assertEquals("Size read must equal size written", blobSize, readCompleteResult.get());
+  }
+
+  /**
+   * Assert that the operation is complete and successful. Note that the future completion and callback invocation
+   * happens outside of the GetOperation, so those are not checked here. But at this point, the operation result should
+   * be ready.
+   */
+  private void assertSuccess(ReadableStreamChannel readableStreamChannel, CountDownLatch readCompleteLatch,
+      AtomicLong readCompleteResult, AtomicReference<Exception> readCompleteException) {
+    try {
+      ByteBufferAsyncWritableChannel asyncWritableChannel = new ByteBufferAsyncWritableChannel();
+      long written;
+      Future<Long> readIntoFuture = readableStreamChannel.readInto(asyncWritableChannel, null);
+      // Compare byte by byte.
+      int readBytes = 0;
+      do {
+        ByteBuffer buf = asyncWritableChannel.getNextChunk();
+        int bufLength = buf.remaining();
+        Assert.assertTrue("total content read should not be greater than length of put content",
+            readBytes + bufLength <= putContent.length);
+        while (buf.hasRemaining()) {
+          Assert.assertEquals("Get and Put blob content should match", putContent[readBytes++], buf.get());
+        }
+        asyncWritableChannel.resolveOldestChunk(null);
+      } while (readBytes < putContent.length);
+      written = readIntoFuture.get();
+      Assert.assertEquals("the returned length in the future should be the length of data written", (long) readBytes,
+          written);
+      Assert.assertNull("There should be no more data in the channel", asyncWritableChannel.getNextChunk(0));
+      readCompleteResult.set(written);
+    } catch (Exception e) {
+      readCompleteException.set(e);
+    } finally {
+      readCompleteLatch.countDown();
+    }
+  }
+
+  /**
+   * Submit all the requests that were handed over by the operation and wait until a response is received for every
+   * one of them.
+   * @param requestList the list containing the requests handed over by the operation.
+   * @return the list of responses from the network client.
+   * @throws IOException
+   */
+  private List<ResponseInfo> sendAndWaitForResponses(List<RequestInfo> requestList)
+      throws IOException {
+    int sendCount = requestList.size();
+    // Shuffle the replicas to introduce randomness in the order in which responses arrive.
+    Collections.shuffle(requestList);
+    List<ResponseInfo> responseList = new ArrayList<>();
+    responseList.addAll(networkClient.sendAndPoll(requestList));
+    requestList.clear();
+    while (responseList.size() < sendCount) {
+      responseList.addAll(networkClient.sendAndPoll(requestList));
```

Comment:
Yeah it could. For the main code, that use case does not arise, as the `OperationController` gets a valid list after the poll and that is the list passed to this method. Will defer doing this for now.